Fix for UTF16 vs. UTF32 positions.

rish987 opened the pull request:
Closes #206 following neovim/neovim#16382, though should be backwards-compatible. Will add some proper tests soon.

Julian commented on lua/lean/_util.lua:
```diff
@@ -218,6 +218,12 @@ function M.position_params_valid(params)
 end
 
+function M.make_position_params()
+  -- UTF-32 is used for general UI features and Lean 3 LSP; UTF16 is used for Lean 4 LSP
```
What kind of general UI features use this? Are they not limited to Lean 3?

I'd love to limit all of this encoding manipulation to just Lean 3 -- and to just having a separate `make_position_params` for Lean 3, rather than carrying around both encodings and having lots of places choose one or the other, that seems easy to make a mistake on.

Julian commented:
@rish987 just want to make sure you're not spinning wheels. Have you seen https://github.com/Julian/lean.nvim/pull/208#discussion_r753808471 ? This PR still seems like it's proceeding with something that can be limited to Lean 3.

rish987 commented:
The way that I see it now is that we have parameters that are used for LSP update requests (stored in whatever format the client needs) and there are parameters for UI features, specifically extmarks, which are in vim's native UTF-8. IMO this is cleaner and more efficient than what we had before, where we ad-hoc converted from whatever client encoding to UTF-8.

rish987 commented:
OK it looks like I've been contending with a bad decision I made way back in #113. I've reworked it to only ever take in UTF-8 params and auto-generate the client position params. Also removed the yucky extra `update()` parameters passed when setting position.

gebner commented on lua/lean/_util.lua:
```diff
@@ -213,11 +213,33 @@ function M.position_params_valid(params)
+  local buf = vim.api.nvim_win_get_buf(0)
+  local row, col = unpack(vim.api.nvim_win_get_cursor(0))
+  row = row - 1
+  local line = vim.api.nvim_buf_get_lines(buf, row, row+1, true)[1]
```
The `vim.api.nvim_buf_line_count` function should be enough here, and doesn't require transfer of any buffer contents. (if it's necessary to check that the line is in range at all).

rish987 replied:
Yea sorry this was dead code, removed.

gebner reviewed (commented):
This is much better now.  AFAICT the `__ui_position_params` field is never sent via LSP, so there's no need to use LSP functions (with encodings) here at all, we could just use the UI rows/columns and pack them like `{ filename = ..., row = ..., col = ... }` (we also always convert the URI directly to a filename, so we might as well store the filename).

I was a bit surprised to see UTF-8 here, but I couldn't find any good documentation on what encoding the various neovim functions expect.  AFAICT we use the UI positions for extmarks and for display.
 1. Extmarks use UTF-8 positions, is that right?
 2. nvim_win_get_cursor is also UTF-8?
 3. For display, I think it would be best to show the window coordinates (which is also what airline does), but I think we could live with UTF-8.

rish987 commented:
Okay reworked to avoid LSP params where possible. Yes vim and Lua are both natively UTF-8, you can assume it anywhere an encoding isn't mentioned.

gebner commented on lua/lean/infoview.lua:
```diff
@@ -592,45 +596,64 @@ function Pin:remove_parent_info(info)
+  end
+
+  local buf_line = vim.api.nvim_buf_get_lines(buf, line, line + 1, false)[1]
+  local end_col = 0
```
I'm not sure I understand the logic here.  Is the code finding the closest valid UTF-8 position?  Does this actually happen (that we get an invalid UTF-8 position)?  If so, please explain this in a comment.

rish987 replied:
Added a comment. This gets the UTF-8 index of the next codepoint.

gebner commented on lua/lean/_util.lua:
```diff
@@ -202,22 +202,38 @@ function M.list_workspace_folders()
+  local row, col = unpack(vim.api.nvim_win_get_cursor(0))
+  row = row - 1
+
+  return { filename = vim.uri_to_fname(vim.uri_from_bufnr(buf)), row = row, col = col }
```
```suggestion
  return { filename = vim.api.nvim_buf_get_name(buf), row = row, col = col }
```

rish987 commented:
Anything else? I think I'll merge this soon since it's backwards compatible anyways and seems like an all-around improvement. Will also mark the new tests as pending until neovim/neovim#16382 is merged.

gebner commented:
The test fails on nightly now.

rish987 commented:
The tests have always been failing, as we're waiting on neovim/neovim#16382. I've locally patched it for our own requests but this is misleading since native LSP requests (hover, goto definition, etc.) will still be broken in 0.6. I could eventually monkey together our own custom LSP commands for Lean 3, but hoping this isn't necessary if neovim/neovim#16382 gets into 0.6.1.
